query/logicalplan/builder.go

```diff
@@ -118,21 +127,86 @@ func (b Builder) Distinct(
 }
 
 func (b Builder) Aggregate(
-	aggExpr []Expr,
+	aggExpr []*AggregationFunction,
 	groupExprs []Expr,
 ) Builder {
+	resolvedAggExpr := make([]*AggregationFunction, 0, len(aggExpr))
+	projectExprs := make([]Expr, 0, len(aggExpr))
+	avgFound := false
+
+	var err error
+	for _, agg := range aggExpr {
+		if agg.Func == AggFuncAvg {
+			avgFound = true
+			sum := &AggregationFunction{
+				Func: AggFuncSum,
+				Expr: agg.Expr,
+			}
+			count := &AggregationFunction{
+				Func: AggFuncCount,
+				Expr: agg.Expr,
+			}
+
+			var (
+				countExpr Expr = count
+				aggType   arrow.DataType
+			)
+			aggType, err = agg.Expr.DataType(b.plan)
+			// intentionally not handling the error here, as it will be handled
+			// in the build function.
+			if !arrow.TypeEqual(aggType, arrow.PrimitiveTypes.Int64) {
+				countExpr = Convert(countExpr, aggType)
```

Comment:
This is the only place where a type conversion is happening implicitly, right? 
This only happens for averaging.

Reply:
Correct, it's just an artifact of average expressions being resolved already by the builder of the logical plan.

Reply:
if value is dynamic float64 field,  below code has problem, because there is no specific dynamic column information in the schema
example:
select avg(value.request) group by name

```go
							t, err := agg.Expr.DataType(v.schema)
							if err != nil {
								v.err = err
								return nil, true
							}

							var typedCount logicalplan.Expr = count
							if !arrow.TypeEqual(t, arrow.PrimitiveTypes.Int64) {
								typedCount = logicalplan.Convert(count, arrow.PrimitiveTypes.Float64)
							}
```

Reply:
BinaryOperation may also require a type conversion to support the following test cases:

```
exec
select value * timestamp as int64, floatvalue * timestamp as float64
----
int64   float64
2       2.000000
12      12.000000
30      30.000000
``` 

```go
case *ast.BinaryOperationExpr:
		// Note that we're resolving exprs as a stack, so the last two
		// expressions are the leaf expressions.
		rightExpr, newExprs := pop(v.exprStack)
		leftExpr, newExprs := pop(newExprs)
		v.exprStack = newExprs

		var frostDBOp logicalplan.Op
		switch expr.Op {
		case opcode.GT:
			frostDBOp = logicalplan.OpGt
		case opcode.LT:
			frostDBOp = logicalplan.OpLt
		case opcode.GE:
			frostDBOp = logicalplan.OpGtEq
		case opcode.LE:
			frostDBOp = logicalplan.OpLtEq
		case opcode.EQ:
			frostDBOp = logicalplan.OpEq
		case opcode.NE:
			frostDBOp = logicalplan.OpNotEq
		case opcode.Plus:
			frostDBOp = logicalplan.OpAdd
		case opcode.Minus:
			frostDBOp = logicalplan.OpSub
		case opcode.Mul:
			frostDBOp = logicalplan.OpMul
		case opcode.Div:
			frostDBOp = logicalplan.OpDiv
		case opcode.LogicAnd:
			v.exprStack = append(v.exprStack, logicalplan.And(leftExpr, rightExpr))
			return nil
		case opcode.LogicOr:
			v.exprStack = append(v.exprStack, logicalplan.Or(leftExpr, rightExpr))
			return nil
		}
		left, _ := leftExpr.DataType(v.schema)
		right, _ := rightExpr.DataType(v.schema)
		if left != nil && right != nil {
			if !arrow.TypeEqual(left, right) {
				if left.ID() == arrow.PrimitiveTypes.Int64.ID() {
					leftExpr = logicalplan.Convert(leftExpr, arrow.PrimitiveTypes.Float64)
				}
				if right.ID() == arrow.PrimitiveTypes.Int64.ID() {
					rightExpr = logicalplan.Convert(rightExpr, arrow.PrimitiveTypes.Float64)
				}
			}
		}

		v.exprStack = append(v.exprStack, &logicalplan.BinaryExpr{
			Left:  leftExpr,
			Op:    frostDBOp,
			Right: rightExpr,
		})
```